yasa/evaluation.py

```diff
@@ -115,123 +109,403 @@ class EpochByEpochEvaluation:
+        >>> acc = ebe.get_agreement().multiply(100).round(0).at["accuracy"]
+        >>> ax.text(0.01, 1, f"Accuracy = {acc}%", ha="left", va="bottom", transform=ax.transAxes)
+
+    When comparing only 2 hypnograms, use the :py:meth:`yasa.Hynogram.evaluate` method:
+
+    >>> hypno_a = yasa.simulate_hypnogram(tib=90, scorer="RaterA", seed=8)
+    >>> hypno_b = hypno_a.simulate_similar(scorer="RaterB", seed=9)
+    >>> ebe = hypno_a.evaluate(hypno_b)
+
+    >>> ebe.get_confusion_matrix()
+    RaterB  WAKE   N1   N2  N3  REM  ART  UNS  Total
+    RaterA
+    WAKE      52   38  126  23   51    0    0    290
+    N1        59    2   27   8   14    0    0    110
+    N2       117   50  105  15   44    0    0    331
+    N3        34   26   62  42   15    0    0    179
+    REM       15   12   13  10    0    0    0     50
+    ART        0    0    0   0    0    0    0      0
+    UNS        0    0    0   0    0    0    0      0
+    Total    277  128  333  98  124    0    0    960
     """
-    def __init__(self, hypno_ref, hypno_test):
-        from yasa.hypno import Hypnogram  # Loading here to avoid circular import
-        assert isinstance(hypno_ref, Hypnogram), "`hypno_ref` must be a YASA Hypnogram"
-        assert isinstance(hypno_test, Hypnogram), "`hypno_test` must be a YASA Hypnogram"
-        assert hypno_ref.n_stages == hypno_test.n_stages, (
-            "`hypno_ref` and `hypno_test` must have the same `n_stages`")
-        if (n_ref := hypno_ref.n_epochs) != (n_test := hypno_test.n_epochs):
-            ## NOTE: would be nice to have a Hypnogram.trim() method for moments like this.
-            if n_ref > n_test:
-                hypno_ref = Hypnogram(hypno_ref.hypno[:n_test], n_stages=hypno_ref.n_stages)
-                n_trimmed = n_ref - n_test
-                warn_msg = f"`hypno_ref` longer than `hypno_test`, trimmed to {n_test} epochs"
-            else:
-                hypno_test = Hypnogram(hypno_test.hypno[:n_ref], n_stages=hypno_test.n_stages)
-                n_trimmed = n_test - n_ref
-                warn_msg = f"`hypno_test` longer than `hypno_ref`, {n_trimmed} epochs trimmed"
-            ## Q: Should be downplayed as INFO?
-            logger.warning(warn_msg)
-        self.hypno_ref = hypno_ref
-        self.hypno_test = hypno_test
-
-    def get_agreement(self):
+
+    def __init__(self, refr_hyps, test_hyps):
+        from yasa.hypno import Hypnogram  # Avoiding circular import
+
+        assert hasattr(refr_hyps, "__iter__"), "`refr_hyps` must be a an iterable"
+        assert hasattr(test_hyps, "__iter__"), "`test_hyps` must be a an iterable"
+        assert type(refr_hyps) == type(test_hyps), "`refr_hyps` and `test_hyps` must be same type"
+        assert len(refr_hyps) == len(
+            test_hyps
+        ), "`refr_hyps` and `test_hyps` must have the same number of hypnograms"
+
+        if isinstance(refr_hyps, dict):
+            # If user provides dictionaries, split into sleep IDs and hypnograms
```

Comment:
We need to provide example on how to go for a long-format dataframe (e.g. with columns "subject", "night", "time", "test", "ref") to two dictionaries:

```python
refr_hyps = {
"subj01_night01": instance of yasa.Hypnogram,
"subj01_night02": instance of yasa.Hypnogram,
...
"subj99_night04": instance of yasa.Hypnogram,
}

test_hyps = {
"subj01_night01": instance of yasa.Hypnogram,
"subj01_night02": instance of yasa.Hypnogram,
...
"subj99_night04": instance of yasa.Hypnogram,
}
```

Reply:
I'm a bit confused on this situation, or at least on handling it. I know that there are many sleep datasets like this (i.e., with multiple sessions from multiple subjects), but wrt evaluation, summary statistics gets weird. Reporting mean and error with variable numbers of sessions from each subject... We would need multiple summary levels, no? Would it be as simple as averaging per-subject first and then across subjects?

Reply:
Yeah the worst situation that I have came across is multiple participants with multiple sessions with multiple devices on both hands for each session. You're totally right that a standard average will not work. We could either use a grand-average (average of average), or maybe make use of weights, e.g. if you have two sessions than the weight assigned to each session is 0.5.

Reply:
Okay I was thinking it would be user's responsibility to handle this, whereas YASA only offers simple averaging (assumes all rows/hypnograms are independent. But maybe you're right that we should offer this flexibility. I'll look into implementing this.

Side question: Should we just use "night" / night_id as the _ultimate_ unique identifier? Mostly this is relevant just for semantics in the docstrings, but I think it's important for clarity. "Nights" isn't always true because of naps, but probably makes the most sense. Alternatively, "run" / run_id would be more BIDS-ey. They use "subject"/"session"/"run" granularity. Then an additional level of separation is the optional "acquisition" / acq_id (this would be for multiple hands, as you say). We could also just adopt that. I think that's my preference maybe? What do you think?

Reply:
Hmmm, I'm not a huge fan to be honest, since this is so far off from the traditional sleep terminology, and I assume most people are unfamiliar with BIDS. I actually only use this standard for MRI data.

But, "night" is not great either since you can have multiple devices for each unique night and therefore it doesn't necessarily reflect the lowest-possible hierarchical level 🤔 

More broadly, do we have to come up with a fixed hierarchical structure/naming pattern within YASA? Can we just let the users define their own key/levels? e.g. for one analysis I have levels "dataset" / "participant", and then for a second analysis I have "participant" / "night" / "device", etc